src/poetry/repositories/pypi_repository.py

```diff
@@ -243,12 +243,16 @@ def _get_release_info(
 
     def _get(self, endpoint: str) -> dict[str, Any] | None:
         try:
-            json_response = self.session.get(self._base_url + endpoint)
+            json_response = self.session.get(
+                self._base_url + endpoint, raise_for_status=False
+            )
         except requests.exceptions.TooManyRedirects:
             # Cache control redirect loop.
             # We try to remove the cache and try again
             self.session.delete_cache(self._base_url + endpoint)
-            json_response = self.session.get(self._base_url + endpoint)
+            json_response = self.session.get(
+                self._base_url + endpoint, raise_for_status=False
+            )
 
         if json_response.status_code == 404:
```

Comment:
Thinking a bit about what @dimbleby said, it is possible that we can mitigate the "bad data" scenario by doing something like this. 


```suggestion
        if not 200 <= json_response.status_code <  300:
```